### PR TITLE
[6.5] UI port subscription tests

### PR DESCRIPTION
**Pull request by ldjebran**
closes https://github.com/SatelliteQE/robottelo/issues/6405

```
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v tests/foreman/ui_airgun/test_subscription.py 
============================================ test session starts =============================================
platform linux -- Python 3.6.6, pytest-3.6.1, py-1.5.4, pluggy-0.6.0 -- /home/dlezz/.pyenv/versions/3.6.6/envs/robottelo/bin/python
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: file
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, services-1.3.0, mock-1.10.0, forked-0.2, cov-2.5.1
collecting 4 items                                                                                           2018-11-20 16:44:01 - conftest - DEBUG - BZ deselect is disabled in settings

collected 4 items                                                                                            

tests/foreman/ui_airgun/test_subscription.py::test_positive_end_to_end PASSED                          [ 25%]
tests/foreman/ui_airgun/test_subscription.py::test_positive_access_with_non_admin_user_without_manifest PASSED [ 50%]
tests/foreman/ui_airgun/test_subscription.py::test_positive_view_vdc_subscription_products PASSED      [ 75%]
tests/foreman/ui_airgun/test_subscription.py::test_positive_view_vdc_guest_subscription_products PASSED [100%]

======================================== 4 passed in 1590.39 seconds =========================================
```
put do not merge to investigate more on one test, and open a bug

### CODE REVIEW

File: robottelo/cli/factory.py [hunk 1]
```diff
@@ -3942,6 +3943,9 @@ def virt_who_hypervisor_config(
     :param bool exec_one_shot: whether to run the virt-who one-shot command
         after startup
     :param bool upload_manifest: whether to upload the organization manifest
+    http://sesame.lab.eng.rdu2.redhat.com/pub/rhel7.repo
+    :param list added_repos: (Optional) a list of repositories dict options to setup additionally.
```
Comment (abalakh) [inline robottelo/cli/factory.py]: i'd vote for `extra_repos` then, or smth else if you have better ideas. 'added repos' sounds like the repos which are already added to me, which is misleading.

File: robottelo/cli/factory.py [hunk 2]
```diff
@@ -3998,6 +3993,10 @@ def virt_who_hypervisor_config(
         ],
         install_katello_agent=False,
     )
+    # configure manually RHEL custom repo url as sync time is very big
+    # (more than 2 hours for RHEL 7Server) and not critical in this context.
+    rhel_repo_url = getattr(settings, 'rhel{0}_repo'.format(DISTROS_MAJOR_VERSION[DISTRO_RHEL7]))
+    virt_who_vm.configure_rhel_repo(rhel_repo_url)
```
Comment (abalakh) [inline robottelo/cli/factory.py]: Shouldn't we cover the case when `rhel*_repo` isn't set in properties?

File: tests/foreman/ui_airgun/test_subscription.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+    user = entities.User(
+        admin=False,
+        role=[role],
+        password=user_password,
+        organization=[org],
+        default_organization=org,
+    ).create()
+    with Session(test_name, user=user.login, password=user_password) as session:
+        session.subscription.navigate_to(session.subscription, 'All')
+        assert not session.browser.url.endswith('katello/403')
```
Comment (abalakh) [inline tests/foreman/ui_airgun/test_subscription.py]: Can we check smth else? Lets say perform some search if searchbox is available on the page with no manifest.
This assert isn't that reliable, user can be redirected to let's say some 'satellite/403', or not redirected anywhere but some text like 'you're not allowed to view this page' can be shown and those would be valid result but our test will fail.

File: tests/foreman/ui_airgun/test_subscription.py [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+    """
+    org = entities.Organization().create()
+    lce = entities.LifecycleEnvironment(organization=org).create()
+    repos_collection = RepositoryCollection(
+        distro=DISTRO_RHEL7, repositories=[RHELAnsibleEngineRepository(cdn=True)])
+    product_name = repos_collection.rh_repos[0].data['product']
+    repos_collection.setup_content(
+        org.id, lce.id, upload_manifest=True, rh_subscriptions=[DEFAULT_SUBSCRIPTION_NAME])
+    with VirtualMachine() as vm:
+        vm.install_katello_ca()
```
Comment (abalakh) [inline tests/foreman/ui_airgun/test_subscription.py]: why not using `setup_virtual_machine` instead?

File: tests/foreman/ui_airgun/test_subscription.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+
+pytestmark = [run_in_one_thread]
+
+if not setting_is_set('fake_manifest'):
+    skip('skipping tests due to missing fake_manifest settings', allow_module_level=True)
+
+
+@tier2
+@upgrade
+def test_positive_end_to_end(session, module_org):
```
Comment (oshtaier) [inline tests/foreman/ui_airgun/test_subscription.py]: you don't need to pass `module_org` here

File: tests/foreman/ui_airgun/test_subscription.py [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+        'subscriptions to hosts and activation keys.',
+        'This action should only be taken in extreme circumstances or for '
+        'debugging purposes.',
+    ]
+    org = entities.Organization().create()
+    _, temporary_local_manifest_path = mkstemp(prefix='manifest-', suffix='.zip')
+    with manifests.clone() as manifest:
+        with open(temporary_local_manifest_path, 'wb') as file_handler:
+            file_handler.write(manifest.content.read())
+
```
Comment (oshtaier) [inline tests/foreman/ui_airgun/test_subscription.py]: redundant line

File: tests/foreman/ui_airgun/test_subscription.py [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+        'Delete all subscriptions that are attached to running hosts.',
+        'Delete all subscriptions attached to activation keys.',
+        'Disable Red Hat Insights.',
+        'Require you to upload the subscription-manifest and re-attach '
+        'subscriptions to hosts and activation keys.',
+        'This action should only be taken in extreme circumstances or for '
+        'debugging purposes.',
+    ]
+    org = entities.Organization().create()
+    _, temporary_local_manifest_path = mkstemp(prefix='manifest-', suffix='.zip')
```
Comment (oshtaier) [inline tests/foreman/ui_airgun/test_subscription.py]: doesn't https://github.com/SatelliteQE/robottelo/blob/master/robottelo/manifests.py#L37 do same job?

File: tests/foreman/ui_airgun/test_subscription.py [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+        with open(temporary_local_manifest_path, 'wb') as file_handler:
+            file_handler.write(manifest.content.read())
+
+    with session:
+        session.organization.select(org.name)
+        session.subscription.add_manifest(temporary_local_manifest_path)
+        assert session.subscription.has_manifest
+        delete_message = session.subscription.read_delete_manifest_message()
+        assert ' '.join(expected_message_lines) == delete_message
+        assert session.subscription.has_manifest
```
Comment (oshtaier) [inline tests/foreman/ui_airgun/test_subscription.py]: you already verified that statement 3 lines above

Comment (ldjebran) [reply tests/foreman/ui_airgun/test_subscription.py]: I have have to very that after cancel I still have my manifest (the read_delete_manifest_message make a cancel), in UI we have a tier1 test for that.

Comment (oshtaier) [reply tests/foreman/ui_airgun/test_subscription.py]: sounds good, but I do not see that you perform cancel operation

File: tests/foreman/ui_airgun/test_subscription.py (hunk 3 above)
Comment (ldjebran) [reply tests/foreman/ui_airgun/test_subscription.py]: we need that assert as that goes along with customer scenario test the next one (we have implemented this test, because of that bug), but also I can add that search to ensure the functionality of the page.

Comment (abalakh) [reply tests/foreman/ui_airgun/test_subscription.py]: Yeah, sure, i meant extra assertion, not removing this one.

File: tests/foreman/ui_airgun/test_subscription.py [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+    user_password = gen_string('alphanumeric')
+    user = entities.User(
+        admin=False,
+        role=[role],
+        password=user_password,
+        organization=[org],
+        default_organization=org,
+    ).create()
+    with Session(test_name, user=user.login, password=user_password) as session:
+        session.subscription.navigate_to(session.subscription, 'All')
```
[the remark this answers is not in the document]
Comment (abalakh) [reply tests/foreman/ui_airgun/test_subscription.py]: I left access to `navigate_to` from tests layer just in case for corner cases when we need only navigation and can't use smth else. But yeah, consider using it a bad tone if some other options are available.

[the remark this answers is not in the document]
Comment (oshtaier) [reply tests/foreman/ui_airgun/test_subscription.py]: yeah, please go ahead with search approach. It will probably return empty table, but that is fine

File: tests/foreman/ui_airgun/test_subscription.py (hunk 7 above)
Comment (ldjebran) [reply tests/foreman/ui_airgun/test_subscription.py]: no that's a linux path used for remote manifest upload, but here we need to save it locally and local can windows , linux or mac ...

File: tests/foreman/ui_airgun/test_subscription.py (hunk 8 above)
Comment (ldjebran) [reply tests/foreman/ui_airgun/test_subscription.py]: read_delete_manifest_message function make that cancel , it read the delete message and click on cancel

File: tests/foreman/ui_airgun/test_subscription.py (hunk 7 above)
Comment (ldjebran) [reply tests/foreman/ui_airgun/test_subscription.py]: Got your question :
it does not save it to a file the function return a file like content , we need to save it, to use it after for the UI 
```
:return: A file-like object (``BytesIO`` on Python 3 and
            ``StringIO`` on Python 2) with the contents of the cloned
            manifest.
```

**Comment by ldjebran** [comment]
@abalakh @oshtaier comments addressed.
```
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v tests/foreman/ui_airgun/test_subscription.py 
============================================ test session starts =============================================
platform linux -- Python 3.6.6, pytest-3.6.1, py-1.5.4, pluggy-0.6.0 -- /home/dlezz/.pyenv/versions/3.6.6/envs/robottelo/bin/python
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: redis
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, services-1.3.0, mock-1.10.0, forked-0.2, cov-2.5.1
collecting 5 items                                                                                           2018-11-23 10:22:08 - conftest - DEBUG - BZ deselect is disabled in settings

collected 5 items                                                                                            

tests/foreman/ui_airgun/test_subscription.py::test_positive_end_to_end PASSED                          [ 20%]
tests/foreman/ui_airgun/test_subscription.py::test_positive_access_with_non_admin_user_without_manifest PASSED [ 40%]
tests/foreman/ui_airgun/test_subscription.py::test_positive_access_with_non_admin_user_with_manifest SKIPPED [ 60%]
tests/foreman/ui_airgun/test_subscription.py::test_positive_view_vdc_subscription_products PASSED      [ 80%]
tests/foreman/ui_airgun/test_subscription.py::test_positive_view_vdc_guest_subscription_products PASSED [100%]

=================================== 4 passed, 1 skipped in 1892.19 seconds ===================================
```

File: tests/foreman/ui_airgun/test_subscription.py (hunk 8 above)
Comment (oshtaier) [reply tests/foreman/ui_airgun/test_subscription.py]: it didn't, but you added that in your airgun PR

File: tests/foreman/ui_airgun/test_subscription.py [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
```
Comment (oshtaier) [inline tests/foreman/ui_airgun/test_subscription.py]: same here
and basically `assert not session.browser.url.endswith('katello/403')` doesn't make sense if you do search afterwards like in previous test case

And I read both 1417082, 1651981 BZ. We do not need to verify `assert not session.browser.url.endswith('katello/403')` explicitly. If search is working for both tests then necessary page showing content.
so for case above do:
        assert not session.subscription.search('')
        assert not session.subscription.has_manifest

and for this:
        assert (session.subscription.search(
            'name = "{0}"'.format(DEFAULT_SUBSCRIPTION_NAME))[0]['Name']
                == DEFAULT_SUBSCRIPTION_NAME)

File: robottelo/cli/factory.py [hunk 11]
```diff
@@ -3942,6 +3943,9 @@ def virt_who_hypervisor_config(
     :param bool exec_one_shot: whether to run the virt-who one-shot command
         after startup
     :param bool upload_manifest: whether to upload the organization manifest
+    http://sesame.lab.eng.rdu2.redhat.com/pub/rhel7.repo
```
Comment (oshtaier) [inline robottelo/cli/factory.py]: Oops, and remove server URL

**Comment by ldjebran** [comment]
@oshtaier @abalakh changes done
